tensorflow/contrib/data/python/ops/dataset_ops.py

```diff
@@ -887,6 +887,10 @@ def __init__(self, tensors):
     """See `Dataset.from_tensor_slices()` for details."""
     super(TensorSliceDataset, self).__init__()
     with ops.name_scope("tensors"):
+      tensors = ops.convert_n_to_tensor_or_indexed_slices(tensors)
```

Comment:
I think this raises a `ValueError` if the `tensors` argument is a single `tf.Tensor`, which is a valid "nested structure" to pass here. (`tf.train.slice_input_producer()` is more restrictive on the types that it accepts.) It would also fail on more complex nested structures, which are handled by the `nest.flatten()` on the lines below.

Perhaps the right thing to do here is to raise a better error than the `IndexError` when the constructor flattens these arguments into four scalars? (The out-of-range is because `Dataset.from_tensor_slices()` requires each of its tensor arguments to have rank > 1, but there's currently no error checking code for that case.) The message could guide users to specify an explicit `tf.constant()` or `tf.convert_to_tensor()`.